# **Heart Disease - Project 2**
### Analyzing the linear relationship between two quantitative variables.

# **Importing Necessary Python Modules**

Python incorporates a variety of open source add-ins called **modules** that add extra features to the basic setup. The name of the modules is after the `import` statement, and the purpose is in a non-code comment after thew hashtag (#).



In [ ]:
import pandas as pd
import numpy as np
import plotly.express as px
from IPython.display import Image
import statsmodels.api as sm


In [ ]:
# Assigns the URL of the image to display to the name 'image_url'.
image_url = 'https://my.clevelandclinic.org/-/scassets/images/org/health/articles/24129-heart-disease-illustration'

# Display the image
Image(url=image_url, width = 600)

# **Context**

Coronary heart disease (CHD), also referred to as coronary artery disease (CAD) involves the reduction of blood flow to the heart muscle due to the build-up of plaque in the arteries of the heart. It is the most common form of cardiovascular disease. Currently, invasive coronary angiography represents the gold standard for establishing the presence, location, and severity of CHD. However, this diagnostic method is costly and associated with morbidity (count of people with the disease) and mortality (count of deaths) in CHD patients. Therefore, it would be beneficial to develop a non-invasive alternative to replace the current gold standard.

Other less invasive diagnostics methods have been proposed in the scientific literature including exercise electrocardiogram, thallium scintigraphy, and fluoroscopy of coronary calcification. However, the diagnostic accuracy of these tests only ranges between 35%-75%. Therefore, it would be beneficial to develop a computer-aided diagnostic tool that could utilize the combined results of these non-invasive tests in conjunction with other patient attributes to boost the diagnostic power of these non-invasive methods with the aim of ultimately replacing the current invasive gold standard.

Three hundred three (303) consecutive patients referred for coronary angiography at the Cleveland Clinic between May 1981 and September 1984 participated in the experiment. No patient had a history of or electrocardiographic evidence of prior myocardial infarction or known valvular or cardiomyopathic diseases.


# **About the Dataset**

The dataset comprises 303 observations, 13 features, and 1 target attribute. A feature is a variable that is believed to contribute to CHD, and is also referred to as a predictive variable. A target variable is the variable you want to predict (CHD, in this situation). The 13 features include the results of the aforementioned non-invasive diagnostic tests along with other relevant patient information. The target variable includes the result of the invasive coronary angiogram which represents the presence or absence of coronary heart disease in the patient. The 14 variables (13 features and 1 target attribute) are described below.

| **Variable**| **Description**                                          |
|:------------|:---------------------------------------------------------|
| AGE         | The age of the individual.                               |
| SEX         | Gender of the individual: 0 = female, 1 = male.          |
| CP          | The type of chest pain experienced by the individual:    |
|             | * 0 = typical angina                                     |
|             | * 1 = atypical angina                                    |
|             | * 2 = non-anginal pain                                   |
|             | * 3 = asymptotic                                         |
| TRESTBPS    | Resting blood pressure of an individual (mmHg)           |
| CHOL        | Serum cholesterol in mg/dL  |
| FBS         | Compares the fasting blood sugar value with 120mg/dL:    |
|             | * 0: fasting blood sugar ≤ 120mg/dL                      |
|             | * 1: fasting blood sugar >120mg/dL                       |
| RESTECG     | Resting electrocardiographic results:                    |
|             | * 0 = normal                                             |
|             | * 1 = having ST-T wave abnormality                       |
|             | * 2 = left ventricular hypertrophy                       |
| THALACH     | Max heart rate achieved, in beats per minute (bpm)       |
| EXANG       | Exercise-induced angina:                                 |
|             | * 0 = No                                                 |
|             | * 1 = Yes                                                |
| OLDPEAK     | ST depression (mm) induced by exercise relative to rest. |
| SLOPE       | Peak exercise ST segment:                                |
|             | * 1 = upsloping                                          |
|             | * 2 = flat                                               |
|             | * 3 = downsloping                                        |
| CA          | Number of major vessels (0-3) colored by fluoroscopy.    |
| THAL        | Thalassemia:                                             |
|             | * 1 = normal                                             |
|             | * 2 = fixed defect                                       |
|             | * 3 = reversible defect                                  |
| TARGET      | Whether the individual is suffering from heart disease:  |
|             | * 0 = absence                                            |
|             | * 1 = present                                            |



Let's take a look at the data. To do this, first we import it directly from the url below.



# **A Snippet of the Data**

Let's take a look at the data. To do this, first we import it directly from the url.

In [ ]:
# Assigns the URL where the data file is stored to 'file_path'.
url='https://raw.githubusercontent.com/ksuaray/STAT108F24_Projects_Jupyter/main/ProjectDataSets/heart.csv'

# Reads in the CSV data file and assigns it to the DataFrame 'df'.
df=pd.read_csv(url)

Next, we can display the data by *typing the name* of the DataFrame. To ensure we can see all columns, we'll use the *pd.set_option* method.

In [ ]:
# Set display options to show all columns
pd.set_option('display.max_columns', None)

# When you type the object name, the object gets printed.
df

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

# **INSTRUCTIONS**

* Use Python to analyze the data set and complete each of the following.
* Replace ellipsis (...) with the relavent names or code.  
* For problems that require a written response, replace the ellipsis (...) by double clicking the text box to start typing.
* Reference the tutorial from activity for assistance.
* Attend office hours if you still need help.

This assignment is intended to explore the univariate linear relationships between quantitative variables in the data. Choose the dependent and independent variables based on your last name.

| Last Name | Dependent Variable (y)  | Independent Variable (x) |
|-----------|-------------------------|--------------------------|
| A-L       | Resting blood pressure  | Serum cholesterol        |
| M-Z       | Max heart rate          | Age                      |




Based on your last name, which variable is explanatory and which is the response?

Explanatory: ...

Response: ...

What do you think the direction (positive/negative) and strength (weak/moderate/strong) will be?

# ...

# **QUESTION 1**
## Scatterplot

**1.1)** Construct a scatterplot that could be used to show the relationship between the variables stated above.

In [ ]:
# Create a scatter plot using Plotly Express

# Scatter plot for A-L
# STUDENTS: replace ... as stated
scatter_plot_AL = px.scatter(df,
                 # Replace ... with the explanatory variable name
                 x='chol',
                 # Replace ... with the response variable name
                 y='trestbps',
                         # Replace ... with a better x-axis label
                 labels={'chol': 'Serum Cholesterol',
                         # Replace ... with a better y-axis label
                         'trestbps': 'Resting Blood Pressure'})

# Updating layout
# STUDENTS: Do not change anything in this block of code.
scatter_plot_AL.update_layout(
    plot_bgcolor='rgba(255,255,255,1)',        # Sets the background color of the plot area to white with full opacity.
    xaxis=dict(
        showline=True,                         # Displays a line on the x-axis.
        showgrid=False,                        # Hides the grid lines on the x-axis.
        linecolor='black'                      # Sets the color of the x-axis line to black.
    ),
    yaxis=dict(
        showline=True,                         # Displays a line on the y-axis.
        showgrid=False,                        # Hides the grid lines on the y-axis.
        linecolor='black'                      # Sets the color of the y-axis line to black.
    ),
    title={
        'text': 'Scatter Plot A-L',   # Sets the title text.
        'y': 0.9,                              # Positions the title 90% of the way up the plot.
        'x': 0.5,                              # Centers the title horizontally.
        'xanchor': 'center',                   # Anchors the title at its center on the x-axis.
        'yanchor': 'top',                      # Anchors the title at the top on the y-axis.
        'font': dict(
            size=16                            # Sets the title font size to 16 (smaller than default).
        ),
    },
    width=575,                                 # Sets the width of the plot.
    height=400                                 # Sets the height of the plot for portrait mode.
)

# Show the plot:
scatter_plot_AL.show()

# Scatter plot for M-Z
# Create a scatter plot using Plotly Express
scatter_plot_MZ = px.scatter(df,
                 x='age', #explanatory variable name
                 y='thalach', #response variable name
                 labels={'age': 'Age', #updates axis labels
                         'thalach': 'Maximum Heart Rate'})

# Updating layout
scatter_plot_MZ.update_layout(
    plot_bgcolor='rgba(255,255,255,1)',        # Sets the background color of the plot area to white with full opacity.
    xaxis=dict(
        showline=True,                         # Displays a line on the x-axis.
        showgrid=False,                        # Hides the grid lines on the x-axis.
        linecolor='black'                      # Sets the color of the x-axis line to black.
    ),
    yaxis=dict(
        showline=True,                         # Displays a line on the y-axis.
        showgrid=False,                        # Hides the grid lines on the y-axis.
        linecolor='black'                      # Sets the color of the y-axis line to black.
    ),
    title={
        'text': 'Scatter Plot M-Z',   # Sets the title text.
        'y': 0.9,                              # Positions the title 90% of the way up the plot.
        'x': 0.5,                              # Centers the title horizontally.
        'xanchor': 'center',                   # Anchors the title at its center on the x-axis.
        'yanchor': 'top',                      # Anchors the title at the top on the y-axis.
        'font': dict(
            size=16                            # Sets the title font size to 16 (smaller than default).
        ),
    },
    width=575,                                 # Sets the width of the plot.
    height=400                                 # Sets the height of the plot for portrait mode.
)

# Show the plot:
scatter_plot_MZ.show()

**1.2)** Describe the relationship between the two variables and include context, and identify the direction, strength, and form. Since you don't yet know the value of correlation (r), estimate what you think the strength is (weak/moderate/strong).




**A-L:** There is a weak positive linear relationship between the resting blood pressure and cholesterol.

**M-Z:** There is a weak negative linear relationship between the max heart rate and age.

**1.3)** Is this what you expected?


# ...

# **QUESTION 2**

## Correlation Coefficient

**2.1)** Calculate the value of the correlation coefficient.

In [ ]:
# Calculate the correlation coefficient

# STUDENTS: Replace ... as stated
# Replace the 1st ... with the response variable name.
# Replace the 2nd ... with the explanatory variable name
correlation_AL = df['trestbps'].corr(df['chol'])
correlation_MZ = df['thalach'].corr(df['age'])

# Print the correlation coefficient
print(f"A-L: Correlation coefficient between chol and resting blood pressure is: r = {correlation_AL}")
print(f"\nM-Z: Correlation coefficient between age and max heart rate is : r = {correlation_MZ}")


A-L: Correlation coefficient between chol and resting blood pressure is: r = 0.12317420653239075

M-Z: Correlation coefficient between age and max heart rate is : r = -0.39852193812106734


**2.2)** Interpret the correlation coefficient.



**A-L:** There is a weak positive linear relationship between the resting blood pressure and cholesterol.

**M-Z:** There is a weak negative linear relationship between the max heart rate and age.

**2.3)** Now that you know the value of correlation, r, did the strength match what you wrote in 1.2?

# ...

# **Question 3**

## Least Squares Regression

**3.1)** Calculate the linear model and add the regression line to the scatterplot.

In [ ]:
# Calculate the LSRL and add the regression line to the scatterplot.

# A-L
# Extract the variables from the dataframe
# STUDENTS: Replace ... as stated below.
# Replace ... with the explantory variable name
X = df['chol']
# Replace ... with the response variable name
Y = df['trestbps']

# Add a constant to the explanatory variable for the regression model.
# This forces Statsmodels to calculate the y-intercept.
# STUDENTS: Do not change anything in this block of code.
X_with_const = sm.add_constant(X)

# Fit the linear regression model (calculate slope and y-intercept)
# STUDENTS: Do not change anything in this block of code.
model_AL = sm.OLS(Y, X_with_const).fit()

# If we repeatedly run this section of code then a new regression line is
# added each run, unless we recreate the scatterplot each iteration.
# STUDENTS: Replace ... as stated below.
scatter_plot_AL = px.scatter(df,
                 # Replace ... with the explanatory variable name
                 x='chol',
                 # Replace ... with the response variable name
                 y='trestbps',
                         # Replace ... with a better x-axis label
                 labels={'chol': 'Serum Cholesterol',
                         # Replace ... with a better y-axis label
                         'trestbps': 'Resting Blood Pressure'})

# Updating layout
# STUDENTS: Do not change anything in this block of code.
scatter_plot_AL.update_layout(
    plot_bgcolor='rgba(255,255,255,1)',        # Sets the background color of the plot area to white with full opacity.
    xaxis=dict(
        showline=True,                         # Displays a line on the x-axis.
        showgrid=False,                        # Hides the grid lines on the x-axis.
        linecolor='black'                      # Sets the color of the x-axis line to black.
    ),
    yaxis=dict(
        showline=True,                         # Displays a line on the y-axis.
        showgrid=False,                        # Hides the grid lines on the y-axis.
        linecolor='black'                      # Sets the color of the y-axis line to black.
    ),
    title={
        'text': 'Scatter Plot A-L',   # Sets the title text.
        'y': 0.9,                              # Positions the title 90% of the way up the plot.
        'x': 0.5,                              # Centers the title horizontally.
        'xanchor': 'center',                   # Anchors the title at its center on the x-axis.
        'yanchor': 'top',                      # Anchors the title at the top on the y-axis.
        'font': dict(
            size=16                            # Sets the title font size to 16 (smaller than default).
        ),
    },
    width=600,                                 # Sets the width of the plot.
    height=400,                                # Sets the height of the plot for portrait mode.
    showlegend=False                           # Disables the legend (key) display
)

# Add the regression line to the scatterplot
scatter_plot_AL.add_scatter(x=X, y=model_AL.predict(X_with_const),
                         mode='lines', name='Regression Line')

# Show the plot
scatter_plot_AL.show()

# M-Z
# Extract the variables from the dataframe
X = df['age']
Y = df['thalach']

# Add a constant to the explanatory variable for the regression model.
# This forces Statsmodels to calculate the y-intercept.
X_with_const = sm.add_constant(X)

# Fit the linear regression model (calculate slope and y-intercept)
model_MZ = sm.OLS(Y, X_with_const).fit()

# If we repeatedly run this section of code then a new regression line is
# added each run, unless we recreate the scatterplot each iteration.
scatter_plot_MZ = px.scatter(df,
                 x='age', #explanatory variable name
                 y='thalach', #response variable name
                 labels={'age': 'Age', #updates axis labels
                         'thalach': 'Maximum Heart Rate'})

# Updating layout: Students do not change anything in this block of code.
scatter_plot_MZ.update_layout(
    plot_bgcolor='rgba(255,255,255,1)',        # Sets the background color of the plot area to white with full opacity.
    xaxis=dict(
        showline=True,                         # Displays a line on the x-axis.
        showgrid=False,                        # Hides the grid lines on the x-axis.
        linecolor='black'                      # Sets the color of the x-axis line to black.
    ),
    yaxis=dict(
        showline=True,                         # Displays a line on the y-axis.
        showgrid=False,                        # Hides the grid lines on the y-axis.
        linecolor='black'                      # Sets the color of the y-axis line to black.
    ),
    title={
        'text': 'Scatter Plot M-Z',   # Sets the title text.
        'y': 0.9,                              # Positions the title 90% of the way up the plot.
        'x': 0.5,                              # Centers the title horizontally.
        'xanchor': 'center',                   # Anchors the title at its center on the x-axis.
        'yanchor': 'top',                      # Anchors the title at the top on the y-axis.
        'font': dict(
            size=16                            # Sets the title font size to 16 (smaller than default).
        ),
    },
    width=600,                                 # Sets the width of the plot.
    height=400,                                # Sets the height of the plot for portrait mode.
    showlegend=False                           # Disables the legend (key) display
)

# Add the regression line to the scatterplot
scatter_plot_MZ.add_scatter(x=X, y=model_MZ.predict(X_with_const),
                         mode='lines', name='Regression Line')

# Show the plot
scatter_plot_MZ.show()

**3.2)** Print the coefficients (Intercept and Slope)

In [ ]:
# Print the coefficients (Intercept and Slope)

# STUDENTS: Do not change the code below.
# The coefficients were calculated when the regression line was added to the
# scatterplot and stored in 'model'. This assigns the values of the y-intercept
# and slope to 'coefficients', then prints the values.

# A-L
coefficients_AL = model_AL.params
coefficients_AL.name='Coefficients'
coefficients_AL

# M-Z
coefficients_MZ = model_MZ.params
coefficients_MZ.name='Coefficients'

# Output the DataFrames
print("A-L Group:")
print(coefficients_AL)
print("\nM-Z Group:")
print(coefficients_MZ)

A-L Group:
const    121.359757
chol       0.041679
Name: Coefficients, dtype: float64

M-Z Group:
const    204.289210
age       -1.005077
Name: Coefficients, dtype: float64


**3.3)** Write the equation of the line that predicts your assigned response variable from your assigned explanatory variable. To type 𝑦̂ you can type “y-hat” or “predicted y”. Do NOT round.

**A-L:** y-hat = 121.359757 + 0.041679x

**M-Z:** y-hat = 204.289210 - 1.005077x

# **QUESTION 4**

## Slope
Interpret the slope

**A-L:** We expect the resting blood pressure to increase by 0.041679 mmHg when the cholesterol increases by 1 mg/dl.

**M-Z:** We expect the max heart rate to decrease by 1.005077 bpm when the age increases by 1 year.

# **QUESTION 5**
## Y-Intercept

Interpret the y-intercept of the least squares regression line in the context of this study. State whether the interpretation is reasonable.

**A-L:** We expect the blood pressure to be 121.359757 mmHg when the cholesterol level is 0 mg/dl.

This does not make sense.

**M-Z:** We expect the max heart rate to be 204.289210 bpm when the age is 0 years old.

This does not make sense.

# **Question 6**
## Prediction

One person in the data set was 74 years old and had a cholesterol level of 269 mg/dL. Predict the value of your dependent variable for this person, using the actual value of the independent variable shown in the table below. Type your work below. Round to 2 decimal places.

| Last Name  | Independent Variable (x) |
|------------|--------------------------|
| **A-L**    | Cholesterol = 269 mg/dL  |
| **M-Z**    | Age = 74 years old       |


**A-L:** y-hat = 121.35975749430595 + 0.04167886407067077(269) = 132.57 bpm

**M-Z:** y-hat = 204.28920953725523 + -1.0050768220596322(74) = 129.91 mg/dL

In [ ]:
# This is NOT in the student notebook.
# A-L prediction
intercept_AL = model_AL.params['const']
slope_AL = model_AL.params['chol']
predicted_AL = (intercept_AL + slope_AL * 269)
print(f"A-L: y-hat = {intercept_AL} + {slope_AL}(269) = {predicted_AL}")

# M-Z prediction
intercept_MZ = model_MZ.params['const']
slope_MZ = model_MZ.params['age']
predicted_MZ = (intercept_MZ + slope_MZ * 74)
print(f"\nM-Z: y-hat = {intercept_MZ} + {slope_MZ}(74) = {predicted_MZ}")


A-L: y-hat = 121.35975749430595 + 0.04167886407067077(269) = 132.5713719293164

M-Z: y-hat = 204.28920953725523 + -1.0050768220596322(74) = 129.91352470484244


# **Question 7**
## Compare
**7.1)** Look up the actual value of your dependent variable for the value of the independent variable above. We know it is for Criminology majors, but be sure to type in the value of the independent variable listed in Question 6 in the code below.

In [ ]:
# Filter the datafram to find the row where your independent variable
#   equals the value from the table in Question 6.

#A-L
# STUDENTS:
# Replace the 1st ... with the explantory variable name
# Replace the 2nd ... with the value from the table in Question 6
# Replace the 3rd ... with the response variable name
actual_dependent_AL = df[df['chol']== 269]['trestbps']
actual_AL = int(actual_dependent_AL.values[0])
print(f"A-L: The actual value of the dependent variable is: {actual_AL}")

#M-Z
actual_dependent_MZ = df[df['age']== 74]['thalach']
actual_MZ = int(actual_dependent_MZ.values[0])
print(f"\nM-Z: The actual value of the dependent variable is: {actual_MZ}")

A-L: The actual value of the dependent variable is: 155

M-Z: The actual value of the dependent variable is: 121


**7.2)** Compare the predicted value of your dependent variable from Question 6 to the actual value from 7.1. Use the actual name of your dependent variable, as modeled in activity.


**A-L:** The actual blood pressure (155 mg/dL) is higher than predicted (132.6 mg/dL).

**M-Z:** The actual max heart rate (121 bpm) is lower than predicted (129.9 bpm).

# **QUESTION 8**

Generate a paragraph of at least 100 words to address one of the following questions. That is, answer only 8a or 8b, but not both.

**8a)** Discuss how analyzing your chosen data set using statistical methods could help you become better prepared for future courses in your major?

...

--OR--

**8b)** Discuss how analyzing your chosen data set using statistical methods could be instrumental in becoming better prepared for your future career?

...


<br><br>
### Once you are done and ready to submit, follow the instructions below to save as a PDF and submit to GradeScope.

### Save as PDF
Note 1: You do not have to select Print Preview. You can print directly from the notebook.
Note 2: Image and graph sizes have been set so you should be able to see them correctly without making any changes to the browser width or the layout (portrait vs landscape).
1. Run all code one last time and make sure your graphs can be seen.
2. File -> Print (or ctrl-p/cmnd-p)
3. Change the "Desination" to PDF.
4. Save the PDF, taking note of where it is saved.

### Submit to GradeScope
**Watch the "GradeScope Submission" video for help.**
1. Login to the Canvas course
2. Click on GradeScope in the course navigation.
3. If you see multiple courses in GradeScope, click on the STAT 108 course
4. Click on the name of the assignment that matches your data set
5. Click on "Submit Work", select PDF
6. Select the PDF you just created
7. You need to tell GradeScope which page each problem answer/output is on. You should see a list of problems on the left, and a display of pages (thumbnails) on the right. Assign pages to questions by clicking on the question number on the left, then clicking on all pages that question is on.
8. After ALL questions have been assigned to their respective page(s), click "Submit"

#### **Still need help? Your STAT 108 team is here to help. Take your laptop to office hours.**
